buildpacks/ruby/src/layers/ruby_install_layer.rs

```diff
@@ -10,51 +27,107 @@ use crate::{
 };
 use commons::gemfile_lock::ResolvedRubyVersion;
 use flate2::read::GzDecoder;
-use libcnb::build::BuildContext;
-use libcnb::data::layer_content_metadata::LayerTypes;
-#[allow(deprecated)]
-use libcnb::layer::{ExistingLayerStrategy, Layer, LayerData, LayerResult, LayerResultBuilder};
 use serde::{Deserialize, Deserializer, Serialize};
 use std::convert::Infallible;
-use std::io;
+use std::io::{self, Stdout};
 use std::path::Path;
 use tar::Archive;
 use tempfile::NamedTempFile;
 use url::Url;
 
-/// # Install Ruby version
-///
-/// ## Layer dir
-///
-/// The compiled Ruby tgz file is downloaded to a temporary directory and exported to `<layer-dir>`.
-/// The tgz already contains a `bin/` directory with a `ruby` executable file.
-///
-/// This layer relies on the CNB lifecycle to add `<layer-dir>/bin` to the PATH.
-///
-/// ## Cache invalidation
-///
-/// When the Ruby version changes, invalidate and re-run.
-///
-pub(crate) struct RubyInstallLayer<'a> {
-    pub(crate) _in_section: &'a dyn SectionLogger, // force the layer to be called within a Section logging context, not necessary but it's safer
-    pub(crate) metadata: RubyInstallLayerMetadata,
+pub(crate) fn handle(
+    context: &libcnb::build::BuildContext<RubyBuildpack>,
+    mut bullet: Print<SubBullet<Stdout>>,
+    metadata: Metadata,
+) -> libcnb::Result<(Print<SubBullet<Stdout>>, LayerEnv), RubyBuildpackError> {
+    let layer_ref = context.cached_layer(
+        layer_name!("ruby"),
+        CachedLayerDefinition {
+            build: true,
+            launch: true,
+            invalid_metadata_action: &|old| match Metadata::try_from_str_migrations(
+                &toml::to_string(old).expect("TOML deserialization of GenericMetadata"),
+            ) {
+                Some(Ok(migrated)) => (
+                    InvalidMetadataAction::ReplaceMetadata(migrated),
+                    "replaced metadata".to_string(),
+                ),
+                Some(Err(error)) => (
+                    InvalidMetadataAction::DeleteLayer,
+                    format!("metadata migration error {error}"),
+                ),
+                None => (
+                    InvalidMetadataAction::DeleteLayer,
+                    "invalid metadata".to_string(),
+                ),
+            },
+            restored_layer_action: &|old: &Metadata, _| {
+                if let Some(diff) = metadata_diff(old, &metadata) {
+                    (
+                        RestoredLayerAction::DeleteLayer,
+                        format!("due to change(s): {}", SentenceList::new(&diff)),
+                    )
+                } else {
+                    (
+                        RestoredLayerAction::KeepLayer,
+                        "using cached version".to_string(),
+                    )
+                }
+            },
```

Comment:
Would it perhaps make sense to change `metadata_diff` to return a `Vec<String>`, and then check if it's empty here instead? Wrapping a vector in `Some` might be a bit redundant/unnecessary.

Just an idea, feel free to disregard!

```suggestion
            restored_layer_action: &|old: &Metadata, _| {
                let diff = metadata_diff(old, &metadata);
                if diff.is_empty() {
                    (
                        RestoredLayerAction::DeleteLayer,
                        format!("due to change(s): {}", SentenceList::new(&diff)),
                    )
                } else {
                    (
                        RestoredLayerAction::KeepLayer,
                        "using cached version".to_string(),
                    )
                }
            },
```

Reply:
FWIW https://github.com/heroku/buildpacks-ruby/blob/6f9ec54b51455fb5f3ff4e5c23768dae285e3855/commons/src/display.rs#L60 displays an empty iterator as "empty" by default so it would show `Clearing cache due to change(s): empty`.

I'll give the unwrapped Vec a try. 

Ultimately I want a declarative interface that could give me both consistency and reduce boilerplate. I've experimented with using the fact that metadata is serializable which means we have access to (some) reflection capabilities at runtime. I'm imagining something like this:

Adding annotations:

```rust
#[derive(MetadataDiff)]
pub(crate) struct MetadataV2 {
    pub(crate) distro_name: String,
    pub(crate) distro_version: String,

    #[diff(name: "CPU architecture")]
    pub(crate) cpu_architecture: String,

    #[diff(name: "Ruby version")]
    pub(crate) ruby_version: ResolvedRubyVersion,

    #[diff(ignore)]
    pub(crate) force_bundle_install_key: String,

    #[diff(ignore)]
    pub(crate) digest: MetadataDigest
}
```

Then a function could take in anything that implements that trait

```rust
fn metadata_diff<T: MetadataDiff>(old: T, now: T) -> DiffIter
```

I'm not planning on shipping that soon, but I wanted to mention where I want to land eventually.

Reply:
I took your suggestion.

Also, despite saying I wouldn't implement the above soon, I realized I could do most of this without a macro (I just need one trait). This allows me to standardize the output in preparation for #326. It also allows me to test cache invalidation logic as a unit test without needing a full integration test.

Reply:
I said so in slack, will repeat here. I've moved those changes to another PR.